Day 5

In [12]:
from functools import reduce
import import_ipynb
import util

sample = "FBFBBFFRLR"



def partition(string, left_key, right_key, r):
    left = r[0]
    right = r[1]
    for c in string:
        if c == left_key:
            right = (right - left) / 2 + left
        elif c == right_key:
            left = right - (right - left) / 2
    return int(left)


def solution(string):
    row = partition(string[:7], "F", "B", [0, 128])
    col = partition(string[7:], "L", "R", [0, 8])
    return row * 8 + col

print(solution("BFFFBBFRRR"))
print(solution("FFFBBBFRRR"))
print(solution("BBFFBBFRLL"))

567
119
820


In [13]:
def find_highest_id():
    ids = [solution(id) for id in util.input_reader('day5.input', 1)]
    print(max(ids))
find_highest_id()

987


In [15]:
def my_seat():
    ids = [solution(id) for id in util.input_reader('day5.input', 1)]
    missing = set()
    for i in range(1024):
        if i not in ids:
            missing.add(i)
    print(missing)
my_seat()

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 603, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999, 1000, 1001, 1002, 1003, 1004, 1005, 1006, 1007, 1008, 1009, 1010, 1011, 1012, 1013, 1014, 1015, 1016, 1017, 1018, 1019, 1020, 1021, 1022, 1023}


603 is the first number that left and right are present